fastlane_bot/tests/deterministic/dtest_token.py

```diff
@@ -0,0 +1,79 @@
+"""
+Token class to store token address and contract object for interacting with the token on the blockchain.
+
+(c) Copyright Bprotocol foundation 2024.
+Licensed under MIT License.
+"""
+from dataclasses import dataclass
+
+from eth_typing import Address
+from web3 import Web3
+from web3.contract import Contract
+
+from fastlane_bot.tests.deterministic.dtest_constants import ETH_ADDRESS
+
+
+@dataclass
+class TestToken:
+    """
+    A class to represent a token on the blockchain.
+
+    Attributes:
+        address: str or Address
+    """
+
+    address: str or Address  # Address after __post_init__, str before
+
+    def __post_init__(self):
+        self.address = Web3.to_checksum_address(self.address)
+        self._contract = None
+
+    @property
+    def contract(self):
+        return self._contract
+
+    @contract.setter
+    def contract(self, contract: Contract):
+        self._contract = contract
+
+    @property
+    def is_eth(self):
+        return self.address == ETH_ADDRESS
+
+
+@dataclass
+class TestTokenBalance:
+    """
+    A class to represent the balance of a token in a wallet in the deterministic tests.
+
+    Attributes:
+        token: TestToken or str
+        balance: int
+    """
+
+    token: TestToken or str  # Token after __post_init__, str before
+    balance: int
+
+    def __post_init__(self):
+        self.token = TestToken(self.token)
+        self.balance = int(self.balance)
+
+    @property
+    def hex_balance(self):
+        return Web3.to_hex(self.balance)
+
+    def faucet_params(self, wallet_address: str = None) -> list:
+        """
+        This method is used to return the faucet parameters for the token balance.
+
+        Args:
+            wallet_address: str
+
+        Returns:
+            list: The faucet parameters.
+        """
+        return (
+            [[self.token.address], self.hex_balance]
+            if self.token.is_eth
+            else [self.token.address, wallet_address, self.hex_balance]
```

Comment:
Why such a different output structure for eth vs for non-eth?
Wouldn't that cause havoc on the caller function?

Reply:
It does not cause havoc, as the implementation fundamentally works as expected. However, to a large extent my efforts here were to **cleanup and improve** upon @NIXBNT 's original solution, while not altering the underlying logic to any great extent - thus, I would defer to Nick on your first question...

Reply:
The return value is immediately used to make a request to Tenderly, as you can see in the following code:

```python
params = token_balance.faucet_params(wallet_address=self.pool_address)
        method_name = RPCEndpoint(
            "tenderly_setBalance"
            if token_balance.token.is_eth
            else "tenderly_setErc20Balance"
        )
        w3.provider.make_request(method=method_name, params=params)
```

Thus, I refer you to the Tenderly documentation for reference, so you can see the expected structure of ETH vs ERC20 requests:

- `tenderly_setBalance` - https://docs.tenderly.co/devnets/advanced/custom-rpc-methods#tenderly_setbalance

- `tenderly_setErc20Balance` - https://docs.tenderly.co/devnets/advanced/custom-rpc-methods#tenderly_seterc20balance

So, while I can see you point of concern, in this case, the reason behind it is strictly 3rd party api's, and IMO changing the handling would add complexity and offer no benefit, and possibly introduce bugs in the process.